File: AiController/src/main/java/com/aimc/controller/LlmClient.java
package com.aimc.controller;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;
import retrofit2.http.POST;

import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * LLM 客户端
 *
 * 封装对 OpenAI Chat Completions API 的调用。
 * 支持: 文本对话、JSON 动作列表提取、多模态视觉决策 (text + image)。
 */
public class LlmClient {

    private final LlmApi api;
    private final Gson gson = new Gson();
    private final String apiKey;
    private final String model;
    private final String baseUrl;
    private final OkHttpClient httpClient;

    public LlmClient(String apiKey, String model) {
        this(apiKey, model, "https://api.openai.com/v1/");
    }

    public LlmClient(String apiKey, String model, String baseUrl) {
        this.apiKey = apiKey;
        this.model = model;
        this.baseUrl = baseUrl;

        OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(chain -> {
                Request request = chain.request().newBuilder()
                    .addHeader("Authorization", "Bearer " + apiKey)
                    .addHeader("Content-Type", "application/json")
                    .build();
                return chain.proceed(request);
            })
            .build();

        this.httpClient = client;

        Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build();

        api = retrofit.create(LlmApi.class);
    }

    private interface LlmApi {
        @POST("chat/completions")
        Call<ChatResponse> chatCompletions(@Body ChatRequest request);
    }

    // ==================== 数据类 ====================

    public static class Message {
        public String role;
        public String content;
        public Message(String role, String content) {
            this.role = role;
            this.content = content;
        }
    }

    public static class ChatRequest {
        public String model;
        public List<Message> messages;
        public double temperature;
        public Integer max_tokens;

        public ChatRequest(String model, List<Message> messages,
                           double temperature, Integer max_tokens) {
            this.model = model;
            this.messages = messages;
            this.temperature = temperature;
            this.max_tokens = max_tokens;
        }
    }

    public static class Choice {
        public Message message;
        public String finish_reason;
    }

    public static class ChatResponse {
        public List<Choice> choices;
    }

    // ==================== 纯文本方法 ====================

    /**
     * 向 LLM 发送纯文本对话请求，返回文本回复
     */
    public String chat(String systemPrompt, String userPrompt) throws Exception {
        List<Message> messages = new ArrayList<>();
        messages.add(new Message("system", systemPrompt));
        messages.add(new Message("user", userPrompt));

        ChatRequest request = new ChatRequest(model, messages, 0.7, 2000);

        Response<ChatResponse> resp = api.chatCompletions(request).execute();
        if (!resp.isSuccessful()) {
            throw new Exception("LLM request failed: " + resp.code());
        }
        ChatResponse response = resp.body();
        if (response == null || response.choices == null || response.choices.isEmpty()) {
            throw new Exception("Empty LLM response");
        }
        return response.choices.get(0).message.content;
    }

    /**
     * 获取结构化的动作列表 (纯文本)
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> getActions(String systemPrompt, String userPrompt) throws Exception {
        String response = chat(systemPrompt, userPrompt);
        String json = extractJsonArray(response);
        if (json == null) {
            throw new Exception("LLM did not return valid JSON array: " + response);
        }
        try {
            List<Map<String, Object>> result = gson.fromJson(json,
                new TypeToken<List<Map<String, Object>>>(){}.getType());
            return result != null ? result : new ArrayList<>();
        } catch (Exception e) {
            throw new Exception("Failed to parse JSON actions: " + e.getMessage(), e);
        }
    }

    // ==================== 多模态视觉方法 ====================

    /**
     * 向 LLM 发送带图片的对话请求 (多模态)
     *
     * 使用 OpenAI Chat Completions API 的 content array 格式:
     * [{"type":"text","text":"..."}, {"type":"image_url","image_url":{"url":"data:..."}}]
     *
     * @param systemPrompt 系统提示词
     * @param userPrompt 用户文本提示
     * @param imageBytes PNG 图片二进制数据
     * @return LLM 文本回复
     */
    public String chatWithImage(String systemPrompt, String userPrompt, byte[] imageBytes) throws Exception {
        JsonObject request = new JsonObject();
        request.addProperty("model", model);
        request.addProperty("temperature", 0.7);
        request.addProperty("max_tokens", 2000);

        JsonArray messages = new JsonArray();

        // System message (text only)
        JsonObject systemMsg = new JsonObject();
        systemMsg.addProperty("role", "system");
        systemMsg.addProperty("content", systemPrompt);
        messages.add(systemMsg);

        // User message (text + image)
        JsonObject userMsg = new JsonObject();
        userMsg.addProperty("role", "user");
        JsonArray content = new JsonArray();

        // Text part
        JsonObject textPart = new JsonObject();
        textPart.addProperty("type", "text");
        textPart.addProperty("text", userPrompt);
        content.add(textPart);

        // Image part
        if (imageBytes != null && imageBytes.length > 0) {
            JsonObject imagePart = new JsonObject();
            imagePart.addProperty("type", "image_url");
            JsonObject imageUrl = new JsonObject();
            imageUrl.addProperty("url", "data:image/png;base64," +
                Base64.getEncoder().encodeToString(imageBytes));
            imagePart.add("image_url", imageUrl);
            content.add(imagePart);
        }

        userMsg.add("content", content);
        messages.add(userMsg);

        request.add("messages", messages);

        // 使用 OkHttp 直接发送请求 (Retrofit 不便处理 content array 格式)
        MediaType jsonType = MediaType.parse("application/json");
        RequestBody body = RequestBody.create(request.toString(), jsonType);
        Request httpRequest = new Request.Builder()
            .url(baseUrl + "chat/completions")
            .post(body)
            .addHeader("Authorization", "Bearer " + apiKey)
            .addHeader("Content-Type", "application/json")
            .build();

        okhttp3.Response response = httpClient.newCall(httpRequest).execute();
        if (!response.isSuccessful()) {
            throw new Exception("LLM request failed: " + response.code());
        }

        String responseStr = response.body().string();
        JsonObject responseJson = JsonParser.parseString(responseStr).getAsJsonObject();
        JsonArray choices = responseJson.getAsJsonArray("choices");
        if (choices == null || choices.isEmpty()) {
            throw new Exception("Empty LLM response");
        }
        return choices.get(0).getAsJsonObject()
            .getAsJsonObject("message")
            .get("content").getAsString();
    }

    /**
     * 获取结构化的动作列表 (带图片)
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> getActionsWithImage(String systemPrompt, String userPrompt, byte[] imageBytes) throws Exception {
        String response = chatWithImage(systemPrompt, userPrompt, imageBytes);
        String json = extractJsonArray(response);
        if (json == null) {
            throw new Exception("LLM did not return valid JSON array: " + response);
        }
        try {
            List<Map<String, Object>> result = gson.fromJson(json,
                new TypeToken<List<Map<String, Object>>>(){}.getType());
            return result != null ? result : new ArrayList<>();
        } catch (Exception e) {
            throw new Exception("Failed to parse JSON actions: " + e.getMessage(), e);
        }
    }

    // ==================== 辅助方法 ====================

    /**
     * 从 LLM 回复中提取 JSON 数组
     * 支持 ```json ... ``` 代码块包裹的格式
     */
    private String extractJsonArray(String text) {
        Pattern jsonBlockRegex = Pattern.compile("```(?:json)?\\s*\\n?([\\s\\S]*?)\\n?```");
        Matcher match = jsonBlockRegex.matcher(text);
        if (match.find()) {
            return match.group(1).trim();
        }

        Pattern arrayRegex = Pattern.compile("(\\[.*\\])", Pattern.DOTALL);
        Matcher arrayMatch = arrayRegex.matcher(text);
        if (arrayMatch.find()) {
            return arrayMatch.group(1).trim();
        }

        return null;
    }
}

File: AiController/src/main/java/com/aimc/controller/LlmClient.kt
package com.aimc.controller

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.*
import retrofit2.*
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * LLM 客户端
 *
 * 封装对 OpenAI Chat Completions API 的调用。
 * 支持通过系统提示词 (System Prompt) 定义 AI 行为。
 *
 * 注意: 此版本为 OpenAI 兼容接口。如果使用其他 LLM (Claude, DeepSeek 等)，
 * 请修改 BASE_URL 和请求格式。
 */
class LlmClient(
    private val apiKey: String,
    private val model: String = "gpt-4o-mini",
    private val baseUrl: String = "https://api.openai.com/v1/"
) {
    private val retrofit: Retrofit
    private val api: LlmApi
    private val gson = Gson()

    init {
        val client = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .addInterceptor { chain ->
                val request = chain.request().newBuilder()
                    .addHeader("Authorization", "Bearer $apiKey")
                    .addHeader("Content-Type", "application/json")
                    .build()
                chain.proceed(request)
            }
            .build()

        retrofit = Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        api = retrofit.create(LlmApi::class.java)
    }

    // ==================== API 接口 ====================

    private interface LlmApi {
        @POST("chat/completions")
        suspend fun chatCompletions(@Body request: ChatRequest): ChatResponse
    }

    // ==================== 请求/响应数据类 ====================

    data class Message(val role: String, val content: String)
    data class ChatRequest(
        val model: String,
        val messages: List<Message>,
        val temperature: Double = 0.7,
        val max_tokens: Int? = 2000
    )

    data class Choice(val message: Message, val finish_reason: String? = null)
    data class ChatResponse(val choices: List<Choice>)

    // ==================== 公开方法 ====================

    /**
     * 向 LLM 发送对话请求，返回文本回复
     */
    suspend fun chat(systemPrompt: String, userPrompt: String): String {
        val request = ChatRequest(
            model = model,
            messages = listOf(
                Message("system", systemPrompt),
                Message("user", userPrompt)
            )
        )
        val response = api.chatCompletions(request)
        return response.choices.firstOrNull()?.message?.content ?: "(no response)"
    }

    /**
     * 获取结构化的动作列表
     *
     * 要求 LLM 返回 JSON 数组格式的动作。
     * 返回解析后的 Map 列表，每个 Map 包含 action 和参数。
     */
    suspend fun getActions(systemPrompt: String, userPrompt: String): List<Map<String, Any>> {
        val response = chat(systemPrompt, userPrompt)

        // 尝试从回复中提取 JSON 数组
        val json = extractJsonArray(response)
            ?: throw Exception("LLM did not return valid JSON array: $response")

        return try {
            val type = object : TypeToken<List<Map<String, Any>>>() {}.type
            gson.fromJson<List<Map<String, Any>>>(json, type) ?: emptyList()
        } catch (e: Exception) {
            throw Exception("Failed to parse JSON actions: ${e.message}", e)
        }
    }

    /**
     * 获取图像 + 文本的多模态回复
     *
     * @param textPrompt 文本提示
     * @param imageBytes PNG 图片字节数组
     */
    suspend fun chatWithImage(textPrompt: String, imageBytes: ByteArray): String {
        // 注意: 多模态需要不同的请求格式 (OpenAI Vision API)
        // 这里提供一个基础实现，实际使用时需要根据 LLM 调整
        val base64Image = android.util.Base64.encodeToString(imageBytes, android.util.Base64.NO_WRAP)

        val content = listOf(
            mapOf("type" to "text", "text" to textPrompt),
            mapOf(
                "type" to "image_url",
                "image_url" to mapOf(
                    "url" to "data:image/png;base64,$base64Image"
                )
            )
        )

        // 此处使用 OkHttp 直接发送请求（因为 Retrofit 接口不同）
        val client = OkHttpClient.Builder()
            .connectTimeout(60, TimeUnit.SECONDS)
            .readTimeout(120, TimeUnit.SECONDS)
            .addInterceptor { chain ->
                val request = chain.request().newBuilder()
                    .addHeader("Authorization", "Bearer $apiKey")
                    .addHeader("Content-Type", "application/json")
                    .build()
                chain.proceed(request)
            }
            .build()

        val jsonBody = gson.toJson(mapOf(
            "model" to model,
            "messages" to listOf(
                mapOf("role" to "user", "content" to content)
            ),
            "max_tokens" to 2000
        ))

        val request = Request.Builder()
            .url(baseUrl + "chat/completions")
            .post(RequestBody.create(MediaType.parse("application/json"), jsonBody))
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw Exception("LLM request failed: ${response.code()} ${response.body?.string()}")
        }

        val respJson = response.body?.string() ?: "{}"
        val respMap = gson.fromJson(respJson, Map::class.java)
        val choices = respMap["choices"] as? List<*>
        val first = choices?.firstOrNull() as? Map<String, Any>
        val message = first?.["message"] as? Map<String, Any>
        return message?.["content"] as? String ?: "(no response)"
    }

    // ==================== 辅助方法 ====================

    /**
     * 从 LLM 回复中提取 JSON 数组
     * 支持 ```json ... ``` 代码块包裹的格式
     */
    private fun extractJsonArray(text: String): String? {
        // 尝试提取代码块中的 JSON
        val jsonBlockRegex = Regex("""```(?:json)?\s*\n?([\s\S]*?)\n?```""")
        val match = jsonBlockRegex.find(text)
        if (match != null) {
            return match.groupValues[1].trim()
        }

        // 尝试直接提取 JSON 数组
        val arrayRegex = Regex("""(\[.*\])""", RegexOption.DOT_MATCH_ALL)
        val arrayMatch = arrayRegex.find(text)
        if (arrayMatch != null) {
            return arrayMatch.groupValues[1].trim()
        }

        return null
    }
}